splicemachine/spliceengine
# DB-11931 Cost model v2

## Short Description
This PR introduces cost model v2.

## Long Description
Cost model v2 contains the following changes:
- `V2ScanCostEstimator` for conglomerate scan cost estimation.
    - local latency set to 3 for OLTP
    - OLAP costs consider number of regions, parallelism, and start overhead
- V2 join cost estimations
    - NLJ separates key join and non-key join with revised formulae
    - MSJ separates OLTP and OLAP with revised formulae.
    - BCJ separates real broadcast join and "broadcast nested loop join" with revised formulae.
    - Merge join is not much touched, only adapted coefficients.
- V2 selectivity estimation
    - Revised input cardinality estimation for inner join.
    - A new join output row count estimator to normalize output row count for a join across all join strategies.
- Many other minor fixes

## How to test
Unfortunately, there is no easy way of testing this change in a complete way. As of this PR is created, all ITs are green on mem platform. Many ITs now also run under v2 cost model and being tests. We try to not adapting any ITs as much as possible.

A small regression suite containing 26 query hashes is tested to make sure that slow queries we've investigated before do not become slow again under v2 cost model.

`JoinCostUnitTests` includes a few tests on join local cost estimation.
